### PR TITLE
[fx/profiler] debug the fx.profiler / add an example test script for fx.profiler

### PR DESCRIPTION
**Pull request by super-dainiu**
### What's new?
I provide an example script for the `fx.profiler` performance test. It is marked as skip.

### Remarks
After I upgraded transformers to the newest version, GPT-2 of hugging face no longer uses `torch.nn.MultiheadAttention` for transformer blocks. So the `torch.nn.functional.softmax` caused some unexpecting problems. They also used `torch.finfo` object, which is weird. I managed to fix them.

### Tests
image

### CODE REVIEW

File: colossalai/fx/profiler/memory.py [hunk 1]
```diff
@@ -73,9 +73,9 @@ def calculate_fwd_tmp(n: Node) -> int:
 
     def is_relu_node(n: Node) -> bool:
         if n.op == 'call_function':
-            return n.target in [torch.nn.functional.relu]
+            return n.target in [torch.nn.functional.relu, torch.nn.functional.softmax]
         elif n.op == 'call_module':
-            return type(n.graph.owning_module.get_submodule(n.target)) in [torch.nn.ReLU]
+            return type(n.graph.owning_module.get_submodule(n.target)) in [torch.nn.ReLU, torch.nn.Softmax]
```
Comment (super-dainiu) [inline colossalai/fx/profiler/memory.py:78]: as I mentioned, softmax is another relu_like node

File: colossalai/fx/profiler/tensor.py [hunk 2]
```diff
@@ -125,5 +125,5 @@ def to(self, *args, **kwargs) -> torch.Tensor:
             device = kwargs['device']
         result = super().to(*args, **kwargs)
         if device is not None:
-            result = MetaTensor(deepcopy(result), fake_device=device)
```
Comment (super-dainiu) [inline colossalai/fx/profiler/tensor.py:128]: No deepcopy required

File: colossalai/fx/profiler/profiler.py [hunk 3]
```diff
@@ -272,7 +272,8 @@ def extract_tensor(x: Any):
             tensor = x._tensor.detach()
             tensor.uuid = x._tensor.uuid
             return tensor
-        return x
+        if not isinstance(x, torch.finfo):
+            return x
```
Comment (super-dainiu) [inline colossalai/fx/profiler/profiler.py:276]: `torch.finfo` is not expected to be stored on `cuda`

File: colossalai/fx/profiler/memory.py (hunk 1 above)
Comment (FrankLeeeee) [reply colossalai/fx/profiler/memory.py:78]: I think this is gonna confuse the future maintainer, what is a relu_like node?

File: tests/test_fx/test_profiler/test_profiler_meta_info_prop.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        # print(f'Memory estimation by saved_tensor_hooks: {fwd_mem / 1024**2}')
+        # =====================================================
+
+        output = gm(data, mask)
+        forward_mem += torch.cuda.memory_allocated(device="cuda:0") / 1024**2 / NUM_STEPS
+        del output
+    return forward_mem, param_mem
+
+
+@pytest.mark.skip("Test for performance, no need for CI")
```
Comment (FrankLeeeee) [inline tests/test_fx/test_profiler/test_profiler_meta_info_prop.py:125]: Mark this with `colossalai.testing.run_on_environment_flag` so that we can still run the test in the future without code change.

File: tests/test_fx/test_profiler/test_profiler_meta_info_prop.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        fwd_flop, bwd_flop = extract_forward_flops(gm)
+        concrete_forward_mem, concrete_param_mem = run_tm_forward(gm)
+
+        print(
+            f'|{m.__name__}|{meta_forward_mem:.3f} MB|{meta_param_mem:.3f} MB|{concrete_forward_mem:.3f} MB|{concrete_param_mem:.3f} MB|fwd_flop={fwd_flop / 1e9:.3f}GFLOPs|bwd_flop={bwd_flop / 1e9:.3f}GFLOPs|'
+        )
+        del model, gm
+
+
+@pytest.mark.skip("Test for performance, no need for CI")
```
Comment (FrankLeeeee) [inline tests/test_fx/test_profiler/test_profiler_meta_info_prop.py:154]: This is the same.

File: colossalai/fx/profiler/memory.py (hunk 1 above)
Comment (super-dainiu) [reply colossalai/fx/profiler/memory.py:78]: I will think about this later. Unfortunately, I am also confused lol. Maybe I should add this to `constants.py`?

File: tests/test_fx/test_profiler/test_profiler_meta_info_prop.py (hunk 4 above)
Comment (super-dainiu) [reply tests/test_fx/test_profiler/test_profiler_meta_info_prop.py:125]: Should it be `colossalai.testing.run_on_environment_flag('fx.profiler')`?

File: colossalai/fx/profiler/memory.py (hunk 1 above)
Comment (FrankLeeeee) [reply colossalai/fx/profiler/memory.py:78]: at least you should define this category of ops so that its semantics is clear.

Comment (super-dainiu) [reply colossalai/fx/profiler/memory.py:78]: Okay, I will make some comments there

File: colossalai/fx/profiler/profiler.py [hunk 6]
```diff
@@ -314,21 +315,17 @@ def get_param_size(x):
         # If there is an argument that this `call_function` is inplace, we should
         # still run the profiling but discard some results regarding `target`
         global do_not_cache
+
         inplace = kwargs.get('inplace', False)
-        if inplace or target in [torch.nn.functional.relu]:
+        if target in RELU_LIKE_OPS:
+            do_not_cache = True
```
Comment (super-dainiu) [inline colossalai/fx/profiler/profiler.py:321]: Here I also did some modification

File: colossalai/fx/profiler/memory.py [hunk 7]
```diff
@@ -71,14 +72,35 @@ def calculate_fwd_tmp(n: Node) -> int:
...
+            _softmax_backward_data_default = torch.ops.aten._softmax_backward_data.default(zeros_like_default, detach_default, None, None);  zeros_like_default = detach_default = None
+            detach_default_1 = torch.ops.aten.detach.default(_softmax_backward_data_default);  _softmax_backward_data_default = None
+            detach_default_2 = torch.ops.aten.detach.default(detach_default_1);  detach_default_1 = None
+
+        Args:
+            n (Node): A node from the graph
+
+        Returns:
+            bool: Whether the node is a ReLU-like node
+        """
```
Comment (super-dainiu) [inline colossalai/fx/profiler/memory.py:96]: A docstring is added

File: colossalai/fx/profiler/constants.py [hunk 8]
```diff
@@ -30,3 +30,10 @@
 CLONE_ATEN = [
     aten.clone.default,
 ]
+
+RELU_LIKE_OPS = [
+    torch.nn.functional.relu,
+    torch.nn.functional.softmax,
+]
```
Comment (FrankLeeeee) [inline colossalai/fx/profiler/constants.py:37]: Still you should define a term for this kind of ops, `relu_like` does not suggest any useful info.

Comment (FrankLeeeee) [reply colossalai/fx/profiler/constants.py:37]: Or call them something like `OUTPUT_SAVED_OPS`?

Comment (super-dainiu) [reply colossalai/fx/profiler/constants.py:37]: I put a docstring in the function which used this constant to illustrate.

File: colossalai/fx/profiler/memory.py [hunk 9]
```diff
@@ -71,14 +74,35 @@ def calculate_fwd_tmp(n: Node) -> int:
         fwd_tmp (int): the result of `fwd_tmp`
     """
 
-    def is_relu_node(n: Node) -> bool:
+    def is_relu_like_node(n: Node) -> bool:
```
Comment (super-dainiu) [inline colossalai/fx/profiler/memory.py:77]: And it's here

File: colossalai/fx/profiler/constants.py (hunk 8 above)
Comment (FrankLeeeee) [reply colossalai/fx/profiler/constants.py:37]: Yup I see it. Just think it could be clearer. I can accept it at the moment but hope it can be improved.

Comment (super-dainiu) [reply colossalai/fx/profiler/constants.py:37]: Okay, I will come back to refactoring the profiler after @Cypher30 finished the backward estimation. I already have some ideas to remove some of these constants without affecting the performance.

**Review by FrankLeeeee — APPROVED** [review]
